### ISSUE TITLE
Non-Intuitive Transport Defaults, Unnecessary Transport Menus, and Unloading Transport Bug

### ISSUE THREAD

**Issue by RuinChristmas** [triplea-game/triplea#553]
The default behavior of transports is the opposite of what a player would want most of the time.  Sometimes the game even makes you click through menus even if there is only one possible option. Also one actual bug where I go through all the menus but still can't unload my transport.

**Setting - three infantry in a land territory, two transports in adjacent sea zone.**

1) click two infantry, move them into the transports.  Game gives you menu box "What transports do you want to load" The default option is two.  Default option should be one. Most times you want to move things with the fewest number of transports.  

2) click three infantry, move them into the transports.  Game gives you the menu box with default option two.  Menu box is unnecessary since I can't move them in one transport.

**Setting - six transports with six infantry, three tanks, three artillery**
(i.e. three transports have inf/art, three have inf/tank) loaded in them, ready to unload.  You see this problem almost every game in revised with UK unloading units into Europe.

3) click a infantry and an arty and move them to the land.  Gives you the menu box "What transports do you want to unload" , but default is take one art from an inf/art tranny and one inf from an inf/tank tranny.  I guess this should be an option but certainly not the default.

3a) unload the inf/arty. From the first menu, change from the default to select 1 inf/art transport.  But then another menu pops up.  "What units do you want to unload"  (But for some reason it shows a picture of transports.)  This menu is unnecessary because I just told you one transport. To make matters worse, the default is two!

3b) [actual bug] now after jumping through all the hoops, I've selected the correct options to unload one infantry and one artillery from one transport into one territory.  I now select one tank and try to move it into a different territory.  But I get an error:  "Transport has already unloaded units into [first territory]"  How can this be?  Even if the game disregarded my selections entirely and insisted on moving an infantry from one of the inf/tank combos, there should still be two tanks available to unload here.

4) Default preference should go to transports that can still move before loading the zero movement transports

5) Default preference should be to use min number of transports with min remaining capacity. Example is on WaW you have T.boat, Transport, and B.transport in the same sea zone and load the following:
1 infantry - defaults to T.boat
2 infantry - defaults to Transport
3-4 infantry - defaults to B.transport
5 infantry - defaults to B.transport and T.boat
6 infantry - default to B.transport and Transport
7 infantry - all

Since I have a problem zipping I'll just upload the save file to sourceforge forum for now. http://tripleadev.1671093.n2.nabble.com/Next-release-1-8-0-11-testers-wanted-tp7590912p7591773.html

**Comment by DanVanAtta** [comment]
If the first five transports that were loaded were all drawn individually, so you would select them by themsevles, would that solve most of the unload problems? I suggest five as a bit of an arbitrary number, eventually we'll need to collapse them together in case someone has 100 transports, one would not want to draw each one individually there.

**Comment by RuinChristmas** [comment]
@DanVanAtta Do you mean load each transport and then unload it (then load/unload the next, etc)? If yes, that's what most players do to get around the problem.  But it gets pretty tedious.

**Comment by ron-murhammer** [comment]
@RuinChristmas Yeah I've run into pretty much all these issues as well. The unloading is the worst one as you click through 2 menus with bad defaults and then it usually still doesn't work right. Unfortunately, these aren't easy to fix from a code standpoint. The logic is rather complex when you have to consider all possibilities but in general it should try to load and unload the fewest transports possible for the given units you've selected.

**Comment by ron-murhammer** [comment]
I looked into this a bit. The method that needs improved for 1&2 is MovePanel.getTransportsToLoad(). This makes call to MoveDelegate.mapTransports() which is where the transport to unit mappings are done but is also used in other places as well. It appears to try to spread units across as many transports as possible instead of minimizing the number of transports used. I think that method needs to be broken into several different methods that handle different loading/unloading scenarios.

**Comment by ron-murhammer** [comment]
Point 1 should be addressed in #647. Planning to tackle the rest of these as well in the near future.

**Comment by ron-murhammer** [comment]
Point 2 should be addressed in #661.

**Comment by DanVanAtta** [comment]
A point 4 to add to the list:
- preference should go to transports that can still move before loading the zero movement transports

**Comment by ron-murhammer** [comment]
Point 4 should be addressed in #666

**Comment by ron-murhammer** [comment]
Point 5 to add to the list:

Default preference should be to use min number of transports with min remaining capacity. Example is on WaW you have T.boat, Transport, and B.transport in the same sea zone and load the following:
- 1 infantry - defaults to T.boat
- 2 infantry - defaults to Transport
- 3-4 infantry - defaults to B.transport
- 5 infantry - defaults to B.transport and T.boat
- 6 infantry - default to B.transport and Transport
- 7 infantry - all

**Comment by ron-murhammer** [comment]
Point 5 should be addressed in #667

**Comment by DanVanAtta** [comment]
Point 3a should be addressed in #817

**Comment by DanVanAtta** [comment]
3b is tricky. We may need to do a large cleanup first before we can actually fix the problem. Basically it appears the problem is here:
```
       line 1017 in MoveValidator
       
        // check whether transport is restricted to another territory
        else if (TransportTracker.isTransportUnloadRestrictedToAnotherTerritory(transport, route.getEnd())) {
          final Territory alreadyUnloadedTo = getTerritoryTransportHasUnloadedTo(undoableMoves, transport);
          for (final Unit unit : TransportTracker.transporting(transport)) {
            result.addDisallowedUnit(TRANSPORT_HAS_ALREADY_UNLOADED_UNITS_TO + alreadyUnloadedTo.getName(), unit);
          }
        }
```
So we append to `disallowedUnit` if any transport has unloaded from the same SZ. Effectively we are not looking up the correct "current" transport. The fix does not appear trivial given the code complexity in this area.

**Comment by ron-murhammer** [comment]
@DanVanAtta Yeah, 3b is going to probably take some time to think about. Not sure if there is gonna be an easy way to fix it without a significant amount of changes.

**Comment by DanVanAtta** [comment]
Moved the last one or two items to new issues (#874, #875). Summarizing the end status here:

1) fixed in PR #647
2) fixed in PR #661
3) moving to issue #875
3a) fixed in #817
3b) moving to issue #/874
4) fixed in PR #666
5) fixed in #667
